src/d3a/models/market/market_structures.py

```diff
@@ -252,8 +275,21 @@ def trade_from_JSON_string(trade_string, current_time):
         trade_dict['residual'] = offer_or_bid_from_JSON_string(trade_dict['residual'],
                                                                current_time)
     trade_dict['time'] = parse(trade_dict['time'])
-    # if 'offer_bid_trade_info' in trade_dict:
-    #     trade_dict['offer_bid_trade_info'] = TradeBidInfo(*trade_dict['offer_bid_trade_info'])
+    if key_in_dict_and_not_none(trade_dict, 'offer_bid_trade_info'):
+        if len(trade_dict['offer_bid_trade_info']) == TradeBidOfferInfo.len():
+            trade_dict['offer_bid_trade_info'] = TradeBidOfferInfo(
+                *trade_dict['offer_bid_trade_info'])
+        elif len(trade_dict['offer_bid_trade_info']) == 3:
+            if _is_bid(trade_string):
+                trade_dict['offer_bid_trade_info'].insert(2, None)
+                trade_dict['offer_bid_trade_info'].insert(3, None)
+                trade_dict['offer_bid_trade_info'] = TradeBidOfferInfo(
+                    *trade_dict['offer_bid_trade_info'])
+            elif _is_offer(trade_string):
+                trade_dict['offer_bid_trade_info'].insert(0, None)
+                trade_dict['offer_bid_trade_info'].insert(1, None)
+                trade_dict['offer_bid_trade_info'] = TradeBidOfferInfo(
+                    *trade_dict['offer_bid_trade_info'])
```

Comment:
Interesting approach, but how about making it a bit more explicit ? In order to avoid this complex deserialization code, the serialization could be simplified and transmit a dict with key-value pairs of the TradeBidOfferInfo (instead of a list that is transmitted now), even including None values. That way the deserialization code will be as simple as `trade_dict['offer_bid_trade_info'] = TradeBidOfferInfo(**trade_dict['offer_bid_trade_info'])` (in theory at least)

Reply:
Attempting to do the fix ass suggested